datajunction-server/datajunction_server/construction/dj_query.py

```diff
@@ -1,206 +1,268 @@
+from datajunction_server.construction.build import build_metric_nodes, build_node
+from datajunction_server.construction.utils import DJErrorException, get_dj_node
+from datajunction_server.models.node import Node, NodeType
+from datajunction_server.sql.parsing import ast
+from datajunction_server.sql.parsing.backends.antlr4 import parse
 from datajunction_server.utils import amenable_name
 
 
 def try_get_dj_node(
     session: Session,
     name: str,
     kinds: Set[NodeType],
-) -> Optional[NodeRevision]:
+) -> Optional[Node]:
     "wraps get dj node to return None if no node is found"
     try:
-        return get_dj_node(session, name, kinds)
+        return get_dj_node(session, name, kinds, current=False)
     except DJErrorException:
         return None
 
 
-def _resolve_metric_nodes(session, col):
+def selects_only_metrics(select: ast.Select) -> bool:
     """
-    Check if a column is a metric and modify the
-    select accordingly
+    Checks that a Select only has FROM metrics
     """
-    joins = []
-    col_name = col.identifier(False)
-    if metric_node := try_get_dj_node(
-        session,
-        col_name,
-        {NodeType.METRIC},
-    ):
-        # if we found a metric node we need to check where it came from
-        parent_select = cast(ast.Select, col.get_nearest_parent_of_type(ast.Select))
-        if not getattr(
-            parent_select,
-            "_validated",
-            False,
-        ):  # pragma: no cover
-            if (
-                len(parent_select.from_.relations) != 1
-                or parent_select.from_.relations[0].primary.alias_or_name.name
-                != "metrics"
-            ):
-                raise DJParseException(
+
+    return (
+        select.from_ is not None
+        and len(select.from_.relations) == 1
+        and len(select.from_.relations[0].extensions) == 0
+        and str(select.from_.relations[0].primary) == "metrics"
+    )
+
+
+def resolve_metric_queries(  # pylint: disable=R0914,R0912,R0915
+    session: Session,
+    tree: ast.Query,
+    ctx: ast.CompileContext,
+    touched_nodes: Set[int],
+    metrics: List[Node],
+):
+    """
+    Find all Metric Node references
+    ensure they belong to a query sourcing `metrics`
+    build the queries as if they were API queries
+    replace them in the original query
+    """
+    for col in tree.find_all(ast.Column):
+        curr_cols = []
+        metric_nodes = []
+        if id(col) in touched_nodes:
+            continue
+        ident = col.identifier(False)
+        if metric_node := try_get_dj_node(session, ident, {NodeType.METRIC}):
+            curr_cols.append((True, col))
+            # if we found a metric node we need to check where it came from
+            parent_select = col.get_nearest_parent_of_type(ast.Select)
+            if not parent_select or not selects_only_metrics(parent_select):
+                raise ast.DJParseException(
```

Comment:
Should this error be more explicit and just say that we can only support queries that reference metrics? In order words, you can't use this query engine to ask DJ for data for a transform node (or am I misinterpreting the logic here?)

Reply:
You can query any nodes actually. To my above example:
```sql
SELECT default.hard_hat.country, default.hard_hat.city  FROM default.hard_hat
```